### ISSUE TITLE
A very slow repository

### ISSUE THREAD

**Issue by magma1447** [borgbackup/borg#2245]
Each host I backup, I backup into 3 different repositories. For one of my hosts, one of the repositories have started to become really slow. I started noticing after I upgraded to beta 3, but I am not sure it's relevant.

I have been running a few tests this evening to narrow the issue down, and show the issue itself. All three repositories are about the same age and should have about the same data, except that one of them is used 3 days per week, while the others are used twice per week. My test case only includes two repositories, one for 2 days and one for 3 days (can be seen on the repository names as well).

Initial setup
-------------

```
cd /tmp
cp /var/log/syslog .
du -hs syslog
3.3M    syslog
```

Client has borgbackup-1.1.0-b3, Server has borgbackup-1.1.0-b3
--------------------------------------------------------------
```
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup create --verbose --compression auto,zlib,3 borg@thor:/data/borg-WedFriSun/pgc-db-01.borg::test_1 syslog
real    1m40.940s
user    0m1.000s
sys     0m0.180s

time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup delete --verbose borg@thor:/data/borg-WedFriSun/pgc-db-01.borg::test_1
real    1m39.064s
user    0m0.948s
sys     0m0.220s
```

Client has borgbackup-1.1.0-b3, Server has borgbackup-1.1.0-b2
--------------------------------------------------------------
```
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup create --verbose --compression auto,zlib,3 borg@thor:/data/borg-WedFriSun/pgc-db-01.borg::test_1 syslog
real    1m39.648s
user    0m1.532s
sys     0m0.216s

time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup delete --verbose borg@thor:/data/borg-WedFriSun/pgc-db-01.borg::test_1
real    1m44.195s
user    0m0.896s
sys     0m0.168s
```

Client has borgbackup-1.1.0-b2, Server has borgbackup-1.1.0-b2
--------------------------------------------------------------
```
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup create --verbose --compression auto,zlib,3 borg@thor:/data/borg-WedFriSun/pgc-db-01.borg::test_1 syslog
real    1m36.910s
user    0m0.964s
sys     0m0.168s

time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup delete --verbose borg@thor:/data/borg-WedFriSun/pgc-db-01.borg::test_1
real    1m37.225s
user    0m1.008s
sys     0m0.160s
```

Switching destination repo
--------------------------
```
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup create --verbose --compression auto,zlib,3 borg@thor:/data/borg-TueSat/pgc-db-01.borg::test_1 syslog
real    0m11.295s
user    0m1.168s
sys     0m0.208s

time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup delete --verbose borg@thor:/data/borg-TueSat/pgc-db-01.borg::test_1
real    0m4.010s
user    0m1.036s
sys     0m0.168s
```

Info of the repos
-----------------
```
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup info borg@thor:/data/borg-WedFriSun/pgc-db-01.borg
Enter passphrase for key ssh://borg@thor/data/borg-WedFriSun/pgc-db-01.borg:
                       Original size      Compressed size    Deduplicated size
All archives:                2.22 TB              1.67 TB            874.72 GB

                       Unique chunks         Total chunks
Chunk index:                  397232               892912

real    0m2.974s
user    0m0.964s
sys     0m0.084s
```
```
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup info borg@thor:/data/borg-TueSat/pgc-db-01.borg
Enter passphrase for key ssh://borg@thor/data/borg-TueSat/pgc-db-01.borg:
                       Original size      Compressed size    Deduplicated size
All archives:                1.93 TB              1.18 TB            630.24 GB

                       Unique chunks         Total chunks
Chunk index:                  373184               757871

real    0m3.569s
user    0m1.292s
sys     0m0.132s
```

My conclusion is that it doesn't matter if I use beta 2 or beta 3. But there is a clear difference in time for creating the backups. For the above test cases the server used 50-100% cpu. It's a i5-3570K. In my real backups it's stuck at 100% for a very long time (hours).

The slower repository is a bit bigger, but doesn't have that many more chunks that I think the speed difference is reasonable.

For my real backups it's mainly an SQL dump. Uncompressed for borg-backup, which compresses it with zlib,3. The dumps are ~12 GB or ~70 GB. The bigger is created ones per day and the smaller ~20 times. Due to the fact that the small one doesn't finish within an hour, the bigger one isn't run on the slow repository.

Backing up the smaller ones takes about 8 minutes for the two good repositories, and somewhere around 5 hours for the slow repository. I haven't clocked it because I haven't been running it manually. But it starts once per hour and I get an email if the pid-file already exists. I have estimated the time by counting the number of emails in a row.

**Comment by magma1447** [comment]
Forgot to say. It's no big deal for me. I can just delete my slow repository and start over on that one. Since I have the other two it's not a big deal. But if the others will grow into this, it's a big problem. But I assume you are interesting in having a debug/test case.

**Comment by ThomasWaldmann** [comment]
Hmm, very strange.

Part of the slowness might come from b3 (we already know it has an issue), but there seems to be another even bigger issue involve here.

Just wild guesses:
- sudden hashtable performance breakdown? (there's another ticket for that)
- is something else using the connection between your client and server so it could have a big impact on the throughput?
- or something running on your client and/or server that puts parallel load on them (cpu/disk/memory)?

**Comment by ThomasWaldmann** [comment]
While the server has a big cpu load in the slow case, can you look at `top`?
Is it all borg who eats cpu or is ssh also very busy?

**Comment by magma1447** [comment]
The process borgbackup uses 100%. I have not seen ssh in the top 10-15.

**Comment by ThomasWaldmann** [comment]
borg serve does not do much computing usually:
- it is mostly I/O, getting/putting chunks from/into segment files.
- it computes crc32 for all segment entries (not that heavy and not expected to vary much)
- it maintains the repo index, which is a big in-memory hashtable (hashtable perf can vary depending on load factor, data, number of tombstones, ...)

**Comment by magma1447** [comment]
I missed a few questions. I should head for bed (3 hours ago).

There doesn't seem to be any memory issues. I doubt it would use 100% cpu if it was swapping at the same time as well. From what I have seen there are 16 GB of  free memory/disk cache.

Regarding something else using the connection I don't really see where you are going. Which connection? The SSH connection? I don't see that it could be anything outside borgbackup doing anything since it's 100% reproducible.

Same regarding load. It's 100% reproducible for me. I do exactly the same thing. I even backup the same file. But I could do another 10-20 tests if anyone is in doubt.

**Comment by magma1447** [comment]
From what I can see. It seems to not use so much cpu in the start. My gut feeling is that it starts using 100% cpu as soon as the data has been transferred. As soon as the server part stops using 100% cpu the whole process is done.

Is the hashtable updated only one time, after the data has been added, and therefore more or less last? If so my gut feeling says it's related.

I could do a new test tomorrow with real data to see that it doesn't start using 100% cpu until at least 5 minutes has passed.

**Comment by ThomasWaldmann** [comment]
Maybe the hash table is in a rather full (near max load factor) or otherwise miserable state (tombstones?), so performance breaks down while cpu load is high and it is busy filling it even more.

When it has reached the max load factor, it will grow the hashtable (== allocate a bigger empty one, transfer all entries from old to new, dealloc old one) and performance will be great again (and thus finishing the rest rather quickly).

You maybe can observe this in memory usage - if memory usage suddenly peaks it is allocating an additional new, bigger hashtable. When it frees the old one memory usage goes down again.

**Comment by ThomasWaldmann** [comment]
@rciorba ^

**Comment by ThomasWaldmann** [comment]
Could you make a backup of the repo index like it was before and after you encountered such a slow behaviour? repodir/index.NNN

**Comment by magma1447** [comment]
-rw------- 1 borg nogroup 30304298 Mar  2 03:09 index.17943-before
-rw------- 1 borg nogroup 30304298 Mar  2 03:11 index.17947-after

Same size before and after. I made a backup of my syslog again. 1.5 minute just as before.

**Comment by magma1447** [comment]
FYI. I have been running recreate on the repository, to recompress the data. I guess it could be relevant as a trigger. This has been done on all of them, but maybe it just caused issues on one.

**Comment by ThomasWaldmann** [comment]
@magma1447 the size include some (more of less) empty space. internally they can be still quite different.
but the HT did not grow, so no resize op happened (or it shrinked again).

So, for these index files you saved, you experienced the cpu eating slowness in between?

If so, can you give me those? They contain sha256 or hmac-sha256 as key and segment number and offset as value, so nothing really private.

**Comment by magma1447** [comment]
Yes, as mentioned, the backup took 1.5 minute. It should take ~10 seconds, so it was a "slow case". I will provide you with links to the files via IRC.

**Comment by enkore** [comment]
Python has a built-in profiler that can be used (but not with static binaries) like this:

borg ... --remote-path "python -m cProfile -o /tmp/output-profile /usr/bin/borg"

**Comment by magma1447** [comment]
After a discussion on IRC yesterday I wanted to try to recreate the index. I created a copy of the repository, deleted the index file, ran a check repair, made a new backup and a delete. Now it's fast again.

Copy of repository, pgc-db-01.copy
----------------------------------
time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup create --verbose --compression auto,zlib,3 borg@thor:/data/borg-WedFriSun/pgc-db-01.copy::test_1 syslog
Enter passphrase for key ssh://borg@thor/data/borg-WedFriSun/pgc-db-01.copy:
Warning: The repository at location ssh://borg@thor/data/borg-WedFriSun/pgc-db-01.copy was previously located at ssh://borg@thor/data/borg-WedFriSun/pgc-db-01.borg
Do you want to continue? [yN] y
real    1m43.326s
user    0m1.044s
sys     0m0.288s

time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup delete --verbose borg@thor:/data/borg-WedFriSun/pgc-db-01.copy::test_1
Enter passphrase for key ssh://borg@thor/data/borg-WedFriSun/pgc-db-01.copy:
Deleting test_1 (1/1):
Archive deleted.
real    1m42.247s
user    0m1.144s
sys     0m0.180s

borg@thor:/data/borg-WedFriSun$ mv pgc-db-01.copy/index.17959 index.17959-copy1

borg@thor:/data/borg-WedFriSun$ time borgbackup check --repair pgc-db-01.copy/
'check --repair' is an experimental feature that might result in data loss.
Type 'YES' if you understand this and want to continue: YES
Enter passphrase for key /data/borg-WedFriSun/pgc-db-01.copy:
real    84m12.317s
user    18m28.868s
sys     9m35.000s

root@pgc-db-01:/tmp# time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup create --verbose --compression auto,zlib,3 borg@thor:/data/borg-WedFriSun/pgc-db-01.copy::test_1 syslog
ynchronizing chunks cache...
Archives: 126, w/ cached Idx: 127, w/ outdated Idx: 5, w/o cached Idx: 4.
Reading cached archive chunk index for small_2017-02-17T22:07:48 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T22:07:17 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T09:07:15 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T00:07:33 ...
Merging into master chunks index ...
Fetching and building archive index for full_2017-02-17T01:39:29 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T20:07:06 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T13:07:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T06:06:17 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T21:07:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T16:07:17 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T23:07:17 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T17:07:23 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T11:07:35 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T07:07:00 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T04:07:09 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T05:12:24 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T10:07:09 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T20:07:10 ...
Merging into master chunks index ...
Reading cached archive chunk index for full_2016-10-16T01:28:19 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T22:08:02 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T13:07:27 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T18:07:14 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T22:09:02 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T21:07:11 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T20:07:11 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T19:06:51 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T18:07:13 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T05:08:26 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T07:07:52 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-26T05:08:07 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T15:07:44 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T23:08:07 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T15:06:57 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T10:07:01 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T05:09:34 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T23:09:31 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T14:07:20 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T16:07:41 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T23:08:52 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T00:06:31 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-26T10:05:51 ...
Merging into master chunks index ...
Reading cached archive chunk index for full_2016-10-09T01:28:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T09:07:59 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T20:08:08 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T16:08:28 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T22:30:09 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T19:07:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T15:06:53 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T04:06:46 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T21:08:02 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T08:12:34 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T05:10:25 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T10:06:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T09:08:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T14:07:28 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T05:07:27 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-24T10:08:50 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T23:08:04 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T21:06:57 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T19:06:37 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T10:07:04 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T19:06:14 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T09:10:54 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-26T20:08:35 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T04:10:25 ...
Merging into master chunks index ...
Reading cached archive chunk index for full_2016-10-12T01:29:41 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T04:11:06 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T18:08:02 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-26T00:06:53 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T11:07:26 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T16:07:49 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T10:06:34 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T17:38:37 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T00:05:54 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T06:07:40 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T17:07:42 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T18:07:39 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T06:08:24 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T12:07:56 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T16:06:45 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T15:08:29 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T11:06:46 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T11:05:53 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T22:07:44 ...
Merging into master chunks index ...
Fetching and building archive index for full_2017-02-15T01:38:39 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T14:07:11 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T17:07:24 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-24T00:08:06 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T08:06:40 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T08:07:00 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T11:07:17 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T15:08:10 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T19:08:32 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-24T05:09:54 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T16:07:37 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T00:07:35 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-03-01T21:08:02 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T12:07:50 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T07:06:48 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T13:05:49 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-26T15:06:49 ...
Merging into master chunks index ...
Fetching and building archive index for full_2017-02-19T01:37:05 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T06:12:24 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T00:07:16 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T05:10:42 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T21:06:55 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T10:07:43 ...
Merging into master chunks index ...
Fetching and building archive index for full_2017-02-22T01:36:51 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T06:07:41 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T17:07:19 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T13:07:55 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T07:05:56 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T07:07:20 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T12:06:05 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T04:09:47 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T14:05:36 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T18:07:22 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T08:09:03 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T09:07:54 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-24T20:05:48 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-15T20:07:18 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-12T12:08:02 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-19T17:06:54 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-24T15:07:16 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-22T08:07:01 ...
Merging into master chunks index ...
Reading cached archive chunk index for small_2017-02-17T14:43:42 ...
Merging into master chunks index ...
Done.
real    0m14.445s
user    0m3.976s
sys     0m0.528s

root@pgc-db-01:/tmp# time BORG_RSH='ssh -i /root/.ssh/thor-borg' borgbackup delete --verbose borg@thor:/data/borg-WedFriSun/pgc-db-01.copy::test_1
Enter passphrase for key ssh://borg@thor/data/borg-WedFriSun/pgc-db-01.copy:
Deleting test_1 (1/1):
Archive deleted.
real    0m5.006s
user    0m1.044s
sys     0m0.200s

**Comment by magma1447** [comment]
Another interesting notice.

After I played around with the index and recreated that, borgbackup refused to use my local cache for the real backups. Since it was too new, fair enough.

I rm -rf'd ~/.cache/borg. After that the real repository is fast as well.

The downside is, I now do not have a reproduction case anymore.

**Comment by ThomasWaldmann** [comment]
I did a very simple analysis of the repo index (NSIndex hashtable), before the slowness happened:

buckets deleted: 360374 used: 397233 empty: 0

The index after it happened, looks very similar.

**No empty buckets. Load ca. 0.53.**

**Comment by ThomasWaldmann** [comment]
likely fixed by #2248 and #2250, see my previous comment.

i am closing this for now. in case the fix does not work, please reopen.

the fix includes getting rid of existing excessive amounts of tombstones (or i.o.w.: no or too few little empty buckets) in on-disk caches/indexes after loading them into memory.

after a successful operation, they are written back to disk and then also the on-disk state is fixed.
